### ISSUE TITLE
Using pre-rendering techniques [$25]

### ISSUE THREAD

**Issue by aggrosoft** [kangax/fabric.js#318]
Currently the _draw implementation of static canvas will always do a full render of each object - no matter what has been modified. The HTML5 Canvas best practices suggest to render an object only once if you have a lot of render passes and then just draw the bitmap representation if needed - this is equal to what flash does using the cacheAsBitmap property and it's display list .

Without caching:

```
// canvas, context are defined
function render() {
  drawMario(context);
  requestAnimationFrame(render);
}
```

Using caching:

```
var m_canvas = document.createElement('canvas');
m_canvas.width = 64;
m_canvas.height = 64;
var m_context = m_canvas.getContext(‘2d’);
drawMario(m_context);

function render() {
  context.drawImage(m_canvas, 0, 0);
  requestAnimationFrame(render);
}
```

For Fabric this would mean that the objects manage a property to tell if a redraw is really needed - this could be bound to the stateProperties. Then the static_canvas will use this property to call render or not. After rendering the bitmap will be cached and reused on the next calls. 

Maybe I just did not inspect the code good enough, is fabric maybe already doing this?

See:

http://www.html5rocks.com/en/tutorials/canvas/performance/
##

**Comment by taikar** [comment]
Is there any news on this request? 

If not, if you any notes/plans on how you were planning on implementing this then I would be happy to contribute to the project!

**Comment by kangax** [comment]
We'll be focusing on performance improvements in the upcoming release. Some things are already in effect — for example, we do itext cursor rendering on a separate canvas. Notes on implementing — hard to tell; these are the usual common techniques used in games, etc. so we just need to adapt them according to fabric's architecture. Any help is welcome of course!

**Comment by jwmao** [comment]
@kangax , any updates regarding the performance improvements using cache. My application is having performance issues when rendering objects on large canvas. Especially on Chrome with hardware acceleration turned on.  There are not a lot of objects on my canvas, my canvas size is 3050px X 3050px. The dragging is very slow and same for mouse clicks. I'm just wondering with caching, it can improve the overall performance.

**Comment by asturur** [comment]
with a 9Mega pixel canvas your problem is probably the use of clearRect() for the full canvas.
you would benefit nearly nothing from object caching, probably more from dirty rects (another possible feature)

**Comment by asturur** [comment]
I opened a PR to address this.
It is in usable state on most shapes, need some testing.
It does not scale yet.
Is a planned feature for 1.7.0, cache will be enabled on by default if it does not affect performance on small quantity of objects.

**Comment by asturur** [comment]
closed with #3317
